File: kakao/2019/kakao2019_7.py
def solution(board):
    answer=0;blocks={}
    for i,r in enumerate(board):
        for j,c in enumerate(r):
            if c:
                if c not in blocks:blocks[c]={'t':51,'b':-1,'l':51,'r':-1,'c':[]}
                blocks[c]['t']=min(i,blocks[c]['t'])
                blocks[c]['b']=max(i,blocks[c]['b'])
                blocks[c]['l']=min(j,blocks[c]['l'])
                blocks[c]['r']=max(j,blocks[c]['r'])
                blocks[c]['c'].append((i,j))
    rkey=1
    while rkey:
        rkey=0
        for blkey,block in blocks.items():
            rmkey=1;z=0
            for i in range(block['t'],block['b']+1):
                for j in range(block['l'],block['r']+1):
                    if board[i][j]==0:
                        z+=1
                        for k in range(i):
                            if board[k][j]:rmkey=0;break
                        if rmkey==0:break
                if rmkey==0:break
            if z==2 and rmkey:
                for i in block['c']:board[i[0]][i[1]]=0
                del blocks[blkey];rkey=1;answer+=1;break
    return answer

File: kakao/2019/kakao2019_7.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(vector<vector<int>> board)
{
    int answer=0,oldanswer;
    int size=board.size(),i,j,k,l,m;
    int diri[5][3]={{1,1,1},{1,2,2},{1,2,2},{1,1,1},{1,1,1}};
    int dirj[5][3]={{0,1,2},{0,0,-1},{0,0,1},{0,-1,-2},{-1,0,1}};
    do 
    {
        oldanswer=answer;
        for(i=0;i<size-1;i++) 
        {
            for(j=0;j<size;j++) 
            {
                if(board[i][j]) 
                {
                    for(k=0;k<5;k++) 
                    {
                        for(l=0;l<3;l++)
                            if((j+dirj[k][l]<0)||(j+dirj[k][l]>=size)||(i+diri[k][l]>=size)||(board[i][j]!=board[i+diri[k][l]][j+dirj[k][l]])) 
                            	break;
                        if(l==3) 
                        {
                            for(l=0;l<3;l++) 
                            {                      
                                for(m=i+diri[k][l]-1;m>=0;m--) 
                                	if(board[m][j+dirj[k][l]]) 
                                		break;
                                if(m>=0&&board[m][j+dirj[k][l]]!=board[i][j]) 
                                	break;
                            }
                            if(l==3)
                            {
                                answer++;
                                board[i][j]=0;
                                for(l=0;l<3;l++) 
                                	board[i+diri[k][l]][j+dirj[k][l]]=0;
                            }
                            break;
                        }
                    }
                }
            }
        }
    }while(answer!=oldanswer);
    return answer;
}